### PR TITLE
fix: add this option to context.ignore

### LINKED ISSUE #4764: Treeshake doesn't work for class with this reference
### Rollup Version

3.7.5

### Operating System (or Browser)

Browser

### Node Version (if applicable)

_No response_

### Link To Reproduction

https://rollupjs.org/repl/?version=3.7.5&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMEZvbyUyMCU3RCUyMGZyb20lMjAnLiUyRnV0aWwuanMnJTNCJTVDbmltcG9ydCUyMCU3QiUyMHVzZWRDb25zdCUyQyUyMGJhciUyMCU3RCUyMGZyb20lMjAnLiUyRnNoYXJlZE1vZHVsZS5qcyclM0IlNUNuJTVDbmNvbnNvbGUubG9nKCdmb28nJTJDJTIwdXNlZENvbnN0KSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMnV0aWwuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwY2xhc3MlMjBGb28lMjAlN0IlNUNuJTIwJTIwJTIwJTIwY29uc3RydWN0b3IoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjB0aGlzLm9wdGlvbjElMjAlM0QlMjAxMjMlM0IlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiU3RCUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJzaGFyZWRNb2R1bGUuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyaW1wb3J0JTIwJTdCJTIwRm9vJTIwJTdEJTIwZnJvbSUyMCcuJTJGdXRpbC5qcyclM0IlNUNuJTVDbiUyRiUyRiUyMFRoZSUyMGZvbGxvd2luZyUyMGxpbmUlMjBzaG91bGQlMjBiZSUyMHJlbW92ZWQuJTVDbmNvbnN0JTIwdW51c2VkQ29uc3QlMjAlM0QlMjBuZXclMjBGb28oKSUzQiU1Q24lNUNuJTJGJTJGJTIwVGhlJTIwZm9sbG93aW5nJTIwbGluZSUyMHNob3VsZCUyMGJlJTIwaW5jbHVkZWQuJTVDbmV4cG9ydCUyMGNvbnN0JTIwdXNlZENvbnN0JTIwJTNEJTIwbmV3JTIwRm9vKCklM0IlNUNuZXhwb3J0JTIwY29uc3QlMjBiYXIlMjAlM0QlMjAxMjMlM0IlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

### Expected Behaviour

Unused variable and its new expression should be removed.

### Actual Behaviour

Rollup still include new expression and drop only unused variable.

### PR DESCRIPTION
**Pull request by TrickyPi**
This PR contains:

- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:
resolves #4764 

### Description

This PR only considers the assignment to the properties of this in the class constructor, and does not work for the following case.
```js
class Main {
	constructor() {
		this.setA();
	}
        setA(){
                this.a = 1;
        }
}
```

### CODE REVIEW

File: src/ast/variables/LocalVariable.ts [hunk 1]
```diff
@@ -167,7 +167,7 @@ export default class LocalVariable extends Variable {
 				);
 			}
 			case INTERACTION_ASSIGNED: {
-				if (this.included) return true;
+				if (this.included && !context.ignore.this) return true;
```
Comment (TrickyPi) [inline src/ast/variables/LocalVariable.ts:170]: @lukastaegert I'm not sure why these tests are failing on CI but succeeding on my local machine, and It seems that all failing tests are related to `sourcemap`, Do you have any idea? By the way, If I remove this change, the tests succeed.

Comment (lukastaegert) [reply src/ast/variables/LocalVariable.ts:170]: No idea, they also appear to be green for me locally. But besides that, I think running this logic for every variable means important side effects will be lost. Look at this example:

https://deploy-preview-4842--rollupjs.netlify.app/repl/?pr=4842&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmxldCUyMHglMjAlM0QlMjAxJTNCJTVDbiU1Q25jbGFzcyUyMEZvbyUyMCU3QiU1Q24lMjAlMjAlMjAlMjBjb25zdHJ1Y3RvcigpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMHglMjAlM0QlMjAyJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMHRoaXMub3B0aW9uMSUyMCUzRCUyMDEyMyUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTdEJTVDbiU1Q25uZXclMjBGb28oKSUzQiU1Q24lNUNuJTJGJTJGJTIwc2hvdWxkJTIwbG9nJTIwJTVDJTIyMiU1QyUyMiU1Q25jb25zb2xlLmxvZyh4KSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE

**Comment by lukastaegert** [comment]
I think this is a good idea, but there are some edge cases this will not handle correctly. E.g. when the constructor first defines a setter:

https://deploy-preview-4842--rollupjs.netlify.app/repl/?pr=4842&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmNsYXNzJTIwRm9vJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGNvbnN0cnVjdG9yKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwT2JqZWN0LmRlZmluZVByb3BlcnR5KHRoaXMlMkMlMjAnb3B0aW9uMSclMkMlMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwc2V0JTIwKHZhbHVlKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBjb25zb2xlLmxvZygnZWZmZWN0JyklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTdEJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCU3RCklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwdGhpcy5vcHRpb24xJTIwJTNEJTIwMTIzJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDbmNvbnN0JTIwdW51c2VkQ29uc3QlMjAlM0QlMjBuZXclMjBGb28oKSUzQiU1Q25jb25zdCUyMHVzZWRDb25zdCUyMCUzRCUyMG5ldyUyMEZvbygpJTNCJTVDbiU1Q25jb25zb2xlLmxvZyh1c2VkQ29uc3QpJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlMkMlMjJuYW1lJTIyJTNBJTIybWFpbi5qcyUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJvdXRwdXQlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiU3RCUyQyUyMnRyZWVzaGFrZSUyMiUzQXRydWUlN0QlN0Q=

For functions, this should be handled the the `replacedVariableInits` logic but apparently, functions have the same problem as classes

https://rollupjs.org/repl/?version=3.7.5&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyZnVuY3Rpb24lMjBGb28oKSUyMCU3QiU1Q24lMjAlMjB0aGlzLm9wdGlvbjElMjAlM0QlMjAxMjMlM0IlNUNuJTdEJTVDbiU1Q25jb25zdCUyMHVudXNlZENvbnN0JTIwJTNEJTIwbmV3JTIwRm9vKCklM0IlNUNuY29uc3QlMjB1c2VkQ29uc3QlMjAlM0QlMjBuZXclMjBGb28oKSUzQiU1Q24lNUNuY29uc29sZS5sb2codXNlZENvbnN0KSUzQiU1Q24lMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlN0QlN0Q=

File: src/ast/variables/LocalVariable.ts (hunk 1 above)
Comment (TrickyPi) [reply src/ast/variables/LocalVariable.ts:170]: Oh, I didn't think about that.

**Comment by TrickyPi** [comment]
Thanks for your review! There are many things that need to be done to complete this pr, it's a bit complicated 😢.

**Comment by lukastaegert** [comment]
I looked at little more into this and pushed a fix to your branch that solves all issues. Even the example with the setter is working, though that was actually broken on master as well.
The idea is that the FunctionNode holds a "constructedEntity" that is re-used and represents the object created when the class or function is called with "new". Whenever "this" is reassigned, it will be tracked on that entity so that e.g. the definition of a setter is not overlooked. Using your new context flag, we ignore mutations to that variable that are not really side effects when checking if the constructor call should be included.
One nice thing is that this handles both class constructors and functions together.

**Comment by TrickyPi** [comment]
You are awesome! I tweaked the test description a bit.

**Review by lukastaegert — APPROVED** [review]

**Pull request merged**